### PR TITLE
SONAR-7267 Do not accept multiple rules with the same key in a repository

### PR DESCRIPTION
**Pull request by simonbrandhof**
SONAR-7267 Do not accept multiple rules with the same key in a repository

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 1]
```diff
@@ -488,14 +488,13 @@ public NewRepositoryImpl setName(@Nullable String s) {
...
+     */
     @Override
     public NewRule createRule(String ruleKey) {
-      if (newRules.containsKey(ruleKey)) {
-        // Should fail in a perfect world, but at the time being the Findbugs plugin
-        // defines several times the rule EC_INCOMPATIBLE_ARRAY_COMPARE
-        // See http://jira.sonarsource.com/browse/SONARJAVA-428
-        Loggers.get(getClass()).warn(format("The rule '%s' of repository '%s' is declared several times", ruleKey, key));
-      }
+      Preconditions.checkArgument(!newRules.containsKey(ruleKey), "The rule '%s' of repository '%s' is declared several times", ruleKey, key);
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: you could do a static import here

File: sonar-plugin-api/src/test/java/org/sonar/api/server/rule/RulesDefinitionTest.java [hunk 2]
```diff
@@ -257,13 +260,12 @@ public void cant_set_blank_repository_name() {
   }
 
   @Test
-  public void warning_if_duplicated_rule_keys() {
+  public void fail_if_duplicated_rule_keys_in_the_same_repository() {
+    expectedException.expect(IllegalArgumentException.class);
```
Comment (sns-seb) [inline sonar-plugin-api/src/test/java/org/sonar/api/server/rule/RulesDefinitionTest.java]: @julienlancelot and I find it a better practice to declare the expectedException right before the statement supposed to throw it
It avoids catching an exception raised by another statement and not have a test failure when we should and it's also better for readability of the UT (it shows where the exception is expected to be raised)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 3]
```diff
@@ -488,14 +488,13 @@ public NewRepositoryImpl setName(@Nullable String s) {
       return this;
     }
 
+    /**
+     * The specified rule key must be unique among the repository. Since version 5.5 it raises
+     * a {@link IllegalArgumentException}, whereas it logs a warning in previous versions.
+     */
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: why not use standard javadoc `@throws`
in addition, shouldn't this be documented on interface method `NewExtendedRepository.createRule(String ruleKey)`?

**Comment by sns-seb** [comment]
some documentation and UT practices comments, otherwise LGTM

**Comment by simonbrandhof** [comment]
@sns-seb feedback applied, thanks

**Pull request closed without merging**